### ISSUE TITLE
Detection of 'Website Blocked' pages used for workspace censorship

### ISSUE THREAD

**Issue by Artoria2e5** [getlantern/lantern#4779]
getlantern/lantern-binaries#4 shows an example of company-level network censorship via HTTP hijacking which uses a 'Website Blocked' to replace the served page. Similar behavior can also be observed in many other workspace censorship solutions, such as Cisco Meraki.

Allowing lantern to match patterns of such pages would extend its scope and help much more users.

**Comment by Artoria2e5** [comment]
Similar code segments actually exist in lantern, for detecting e.g. the Iranian iframe blocking. I think I can start by adding a few patterns for my own use, e.g. `<p class="type_style1"><br>\n      This website is blocked by your network operator.\n      <table style="margin-top:1em" width="65%" border="0" cellpadding="0" cellspacing="0">`.

**Comment by Artoria2e5** [comment]
Additional example: #5267, warning.or.kr shown *through* lantern.

**Comment by myleshorton** [comment]
Interesting @Artoria2e5. Are you personally somewhere that resided behind Meraki or something similar? Would you feel comfortable sharing where if so? This is actually a use case we're extremely interested in, and we'd be happy to pursue this further. FYI @WolfLantern

**Comment by Artoria2e5** [comment]
I am personally resided behind Meraki at school. The router uses Squid, 302s away HTTP, and RSTs off HTTPS. A reason for me to break the rules is to access SourceForge and PUBMED.

**Comment by myleshorton** [comment]
Wow -- they don't let you access SourceForge @Artoria2e5? Is it a large school?

We'd love to dive into how it's working actually. Do you know specifically which box they're using? We'll probably just end up buying the latest version and seeing how we can route around it.

**Comment by Artoria2e5** [comment]
> they don't let you access SourceForge

They do allow GitHub though. Sounds more like glitches in the blacklist, whatever.

> Is it a large school? 

No.

> Do you know specifically which box they're using? 

I think I can check out the AP in the corridor, but I doubt if it is worth the risk as the management/security features listed for [Meraki wireless APs](https://www.meraki.com/products/wireless) are pretty much the same. Note the APs in "wireless" don't have content filtering by themselves, but the all-in-a-box things in "[security appliances](https://www.meraki.com/products/appliances/mx65w)" do. My school can't be using the all-in-a-box solution as these boxes don't support 802.1x, but I guess as long as content filtering feature is present you can test it and generalize it to all Meraki products.

**Comment by myleshorton** [comment]
OK thanks @Artoria2e5!

**Comment by p3rcya** [comment]
I'm more concerned about national level filtering as seen from the warning.or.kr but I think it's the same concept in this issue

**Comment by Artoria2e5** [comment]
Woah, they changed the pattern.

```
$ curl -i -A 'Mozilla/5.0 (whatever, like whatever)' http://store.steampowered.com/app/351640/
HTTP/1.1 302 Found
Location: http://ap.meraki.com/blocked.cgi?blocked_server=104.88.9.113:80&blocked_url=http%3A%2F%2Fstore.steampowered.com%2Fapp%2F351640%2F
Content-Type: text/html
Content-Length: 200
Cache-Control: no-cache
Content-Type: text/html
Pragma: no-cache
Expires: 0
Continue: close

<html><body>You are being <a href='http://ap.meraki.com/blocked.cgi?blocked_server=104.88.9.113:80&amp;blocked_url=http%3A%2F%2Fstore.steampowered.com%2Fapp%2F351640%2F'>redirected</a>.</body></html>
```

**Comment by myleshorton** [comment]
Yeah interesting @Artoria2e5. That's definitely very easy to detect. FYI @fffw.

**Comment by joesis** [comment]
Yeah, they can be handled case by case, whether the redirection is deployed national wide, by ISP, or by any organization.

While more and more sites are moving away from plain HTTP, the hijacking will become less useful. We can even integrate [https-everywhere rules](https://github.com/EFForg/https-everywhere/tree/master/src/chrome/content/rules) to accelerate the process.

**Comment by p3rcya** [comment]
@joesis I think the blocking by the Korean gov is DNS hijacking. So HTTPS won't help

**Comment by joesis** [comment]
@p3rcya The Korean blocking should be in two phase.

1. Resolve blocked sites to a fixed IP address.
2. Host a simple site on the IP address to redirect HTTP to warning.or.kr.

For HTTPS, accessing the IP address would be reset (if no https port is open) or show an error (invalid certificate).

**Comment by p3rcya** [comment]
@oxtoacart The most urgent hijacking is this type https://github.com/getlantern/lantern-internal/issues/198

**Comment by oxtoacart** [comment]
@p3rcya A generalized articulation of this, which will also be useful if/when we reintroduce P2P, is for Lantern proxies to advertise what domains are blocked so that their clients can request those sites from a different proxy.

**Comment by p3rcya** [comment]
@oxtoacart That's a great idea. And the local Lantern client can detect HTTP/DNS hijacking locally, together with a country wide restriction list. 

But for the most imminent fix, could you ask this particular DC to lift such restrictions, if possible? This is implemented by the Korean gov so I don't know if they can have an exemption for all of our IP

**Comment by oxtoacart** [comment]
@p3rcya I've sent an inquiry to tech support for the DC.

**Comment by oxtoacart** [comment]
Since our India proxy server is in India, split-tunneling isn't as important there and one can just choose to "Proxy All" from Advanced Settings which bypasses the blocking auto-detection.
